azurelinuxagent/ga/exthandlers.py

```diff
@@ -310,11 +309,68 @@ def handle_ext_handlers(self, etag=None):
                 logger.info("Extension handling is on hold")
                 return
 
+        dep_level = None
+        deps_wait_until = None
+        handlers_wait_until = datetime.datetime.utcnow()
+        prev_handler = None
+
         self.ext_handlers.extHandlers.sort(key=operator.methodcaller('sort_key'))
         for ext_handler in self.ext_handlers.extHandlers:
-            # TODO: handle install in sequence, enable in parallel
+            # Use updated time limit for each dependency level
+            if dep_level != ext_handler.sort_key():
+                dep_level = ext_handler.sort_key()
+                deps_wait_until = handlers_wait_until
+                handlers_wait_until += datetime.timedelta(minutes=DEFAULT_EXT_TIMEOUT_MINUTES)
+
+            # Make sure that the previous extension is installed.
+            # If installed successfully, proceed with installing the current handler.
+            # Otherwise, skip the rest of them.
+            if dep_level >= 0:
+                if self.wait_for_prev_handler_installation(prev_handler, ext_handler, deps_wait_until):
+                    prev_handler = ext_handler
+                else:
+                    break
+
             self.handle_ext_handler(ext_handler, etag)
-    
+
+    def wait_for_prev_handler_installation(self, prev_handler, cur_handler, wait_until):
+        '''
+        Check the status of the previous extension installed.
+        Wait until it becomes success or times out.
+        Return True if it is installed successfully. False if timed out.
+        '''
+        # No need to wait on anything for the very first extension
+        if prev_handler == None:
+            return True
+
+        handler_i = ExtHandlerInstance(prev_handler, self.protocol)
+        for ext in prev_handler.properties.extensions:
+            success_status, status = handler_i.is_ext_status_success(ext)
+
+            # Keep polling for the extension status until it becomes success or times out
+            while not success_status and datetime.datetime.utcnow() <= wait_until:
+                time.sleep(5)
+                success_status, status = handler_i.is_ext_status_success(ext)
+
+            # In case of timeout, we log it and return false so that the extensions waiting
+            # on this one can be skipped processing
+            if not success_status:
+                msg = "Timeout waiting for success status " \
+                      "from dependency {0}/{1} for {2} " \
```

Comment:
I'd remove the "dependency" part from this message, since there may not be any dependency between cur_handler and prev_handler